### PR TITLE
Use a configurable location to load global Devtools settings

### PR DESCRIPTION
**Pull request by scottmf**
This allows separate projects to keep their own settings where common settings
such as spring.* or server.* don't conflict.

https://github.com/spring-projects/spring-boot/issues/17923

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@scottmf Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=17924)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=17924) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@scottmf Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=17924)!

File: spring-boot-project/spring-boot-devtools/src/main/java/org/springframework/boot/devtools/env/DevToolsHomePropertiesPostProcessor.java [hunk 1]
```diff
@@ -61,6 +61,17 @@ public void postProcessEnvironment(ConfigurableEnvironment environment, SpringAp
 		}
 	}
 
+	protected File getProjectRootFolder() {
```
Comment (chenqimiao) [inline spring-boot-project/spring-boot-devtools/src/main/java/org/springframework/boot/devtools/env/DevToolsHomePropertiesPostProcessor.java]: It seems a breaking change.
When `PropertyFile` of `HomeFolder` have a property which not exist in `ProjectRootFolder`, after this change the property will be invalid.

**Review by sebasanjose — APPROVED** [review]

File: spring-boot-project/spring-boot-devtools/src/main/java/org/springframework/boot/devtools/env/DevToolsHomePropertiesPostProcessor.java (hunk 1 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot-devtools/src/main/java/org/springframework/boot/devtools/env/DevToolsHomePropertiesPostProcessor.java:64]: I don't think this is a breaking change as you have to opt in to looking in the new location by setting that system property or environment variable.

**Comment by wilkinsona** [comment]
While not a breaking change, [this comment](https://github.com/spring-projects/spring-boot/pull/17924#discussion_r356950946) raises the question of whether we should use the user's home directory to create property sources with lower precedence than those created for the custom home directory. A benefit of this is that it would allow project-specific settings to be set in the custom location and for common settings to be set in the user's home directory. A downside is that it would then be impossible for a property in the user's home directory to be unset. I'm leaning towards keeping things simple and only using one location as in [this branch](https://github.com/wilkinsona/spring-boot/tree/gh-17924). Flagging for team attention to set what everyone else thinks.

**Comment by wilkinsona** [comment]
We discussed this today and decided not to use the default global location as a lower-precedence property source when a custom global location is configured.

**Comment by wilkinsona** [comment]
@scottmf Thanks very much for your patience and for making your first contribution to Spring Boot.

**Pull request closed without merging**